### PR TITLE
WIP: bpo-34990: year 2038 problem in compileall.py

### PR DESCRIPTION
**Pull request by matrixise**
Use 'L' as an unsigned long for the timestamp

https://bugs.python.org/issue34990

### CODE REVIEW

**Comment by bmwiedemann** [comment]
`Lib/test/test_compileall.py` also contains 2 instances of `<4sll` that probably should be updated accordingly.
and you could add the test code provided by xtreak in the bug.

**Comment by matrixise** [comment]
@bmwiedemann yep, I was in the plane from PySS and I just commited my code, and now, I just saw the test of @tirkarthi . I am going to include it.

**Comment by tirkarthi** [comment]
@matrixise Thanks for the PR. I think it's a NEWS entry worthy fix to add a short note that `compileall` now supports files with modification time greater than or equal to 19-01-2038 (I guess)

**Comment by tirkarthi** [comment]
@matrixise  Sorry, I just noticed you have added one :)

**Comment by matrixise** [comment]
@tirkarthi please, could you check my PR with the last commits.

**Comment by bmwiedemann** [comment]
btw: the first bad mtime is 0x80000000 or 2147483648 or 2038-01-19T03:14:08+00:00

File: Lib/test/test_compileall.py
Comment (bmwiedemann) [inline Lib/test/test_compileall.py:201]: not sure, but maybe this is not a good place in this file to add the test in, because SOURCE_DATE_EPOCH changes the default .pyc type to be hash-based, which would not hit the error condition. See #9607

**Comment by tirkarthi** [comment]
@matrixise Thanks, Looks good to me from the context of the issue reported but I have limited understanding about the code so I hope someone else might have a better look. Seems there was some discussion about handling 2038 problem as part of the original PR where this code was added : https://github.com/python/cpython/pull/4575#discussion_r153376173

File: Lib/test/test_compileall.py
Comment (matrixise) [reply Lib/test/test_compileall.py:201]: I am not sure, just copy the code of @tirkarthi . I will wait for a review from @vstinner or an other core-dev.

**Comment by matrixise** [comment]
This is the reason why I move to 'L' instead of 'l'

File: Lib/compileall.py
Comment (serhiy-storchaka) [inline Lib/compileall.py:143]: `mtime & 0xFFFFFFFF`

**Review by serhiy-storchaka — CHANGES_REQUESTED** [review]
Check also other parts of the stdlib that work with mtime in pyc files. For example zipimport and checkpyc.py need updates. Maybe there are other parts.

File: Lib/compileall.py
Comment (serhiy-storchaka) [inline Lib/compileall.py]: Also, it would be better to handle the second field as an unsigned integer too.

Comment (vstinner) [reply Lib/compileall.py]: importlib uses int(mtime) & 0xFFFFFFFF, I suggest... int(mtime) & 0xFFFF_FFFF for readability :-D

**Comment by matrixise** [comment]
@serhiy-storchaka last update for `Tools/scripts/cherrypyc.py` by in June 2013, we only read the magic number and timestamp, just that: https://github.com/python/cpython/blob/master/Tools/scripts/checkpyc.py#L45-L46

In https://github.com/python/cpython/blob/master/Lib/zipimport.py#L602 you convert the dword as a uint32 and LGTM.

**Comment by matrixise** [comment]
@vstinner and @serhiy-storchaka I have fixed mtime with 0xFFFF_FFFF and use an unsigned long, where I found `<4slL` but for the other files, I am not an expert of the API and I don't know where I could search, `zipimport` and `checkpy.py` seem to be good.

File: Tools/scripts/checkpyc.py
Comment (serhiy-storchaka) [inline Tools/scripts/checkpyc.py:55]: mtime can't be -1 after adding `& 0xFFFF_FFFF`.

File: Lib/test/test_zipimport.py
Comment (serhiy-storchaka) [inline Lib/test/test_zipimport.py:42]: I think the above few lines can be removed.

File: Lib/test/test_zipimport.py
Comment (serhiy-storchaka) [inline Lib/test/test_zipimport.py:43]: Signed format units are used.

File: Tools/scripts/checkpyc.py
Comment (serhiy-storchaka) [inline Tools/scripts/checkpyc.py]: Sees the definition of get_long(). It returns an unsigned 32-bit integer, and returns -1 only on error. You need to convert not it, but the file timestamp.

Comment (bmwiedemann) [inline Tools/scripts/checkpyc.py]: This is actually a NOP, because `mtime_str` is just 4 bytes, that cannot encode anything larger than `0xFFFF_FFFF`

**Review by serhiy-storchaka — CHANGES_REQUESTED** [review]
zipimport needs truncating the timestamp to 32 bit for comparison.

Also update _code_to_timestamp_pyc() in _bootstrap_external.

File: Tools/scripts/checkpyc.py
Comment (matrixise) [reply Tools/scripts/checkpyc.py]: thanks, I am not familiar with the operations on the bits.

File: Tools/scripts/checkpyc.py
Comment (matrixise) [reply Tools/scripts/checkpyc.py:55]: maybe add the 0xFFFF_FFFF on the line 57?

File: Lib/test/test_zipimport.py
Comment (matrixise) [reply Lib/test/test_zipimport.py:42]: the lines from 38->41?

File: Lib/test/test_zipimport.py
Comment (bmwiedemann) [inline Lib/test/test_zipimport.py:43]: why this revert?

Comment (matrixise) [reply Lib/test/test_zipimport.py:43]: just because @serhiy-storchaka indicated that the signed was used here.

File: Lib/importlib/_bootstrap_external.py
Comment (vstinner) [inline Lib/importlib/_bootstrap_external.py]: This change is useless, _pack_uint32() already does "& 0xFFFFFFFF". importlib can be left unchanged.

File: Lib/test/test_compileall.py
Comment (vstinner) [inline Lib/test/test_compileall.py]: That's why I like the underscore: it's easier to spot bugs. You missed something :-)

File: Lib/test/test_compileall.py
Comment (vstinner) [reply Lib/test/test_compileall.py:201]: It should be moved to CompileallTestsBase.

File: Misc/NEWS.d/next/Core and Builtins/2018-10-15-20-41-18.bpo-34990.NVDjdP.rst
Comment (vstinner) [inline Misc/NEWS.d/next/Core and Builtins/2018-10-15-20-41-18.bpo-34990.NVDjdP.rst]: You describe the fix, but you should describe the fixed bug instead: compileall now supports timestamps newer than year 2038.

File: Tools/scripts/checkpyc.py
Comment (vstinner) [inline Tools/scripts/checkpyc.py:57]: I don't know if this code should be modified, but it is modified: should we also cast to int()?

**Review by vstinner — CHANGES_REQUESTED** [review]

File: Lib/test/test_compileall.py
Comment (matrixise) [reply Lib/test/test_compileall.py:201]: done

File: Tools/scripts/checkpyc.py
Comment (matrixise) [reply Tools/scripts/checkpyc.py:57]: like that?
```python
elif int(mtime) != int(st[ST_MTIME] & 0xFFFF_FFFF)
```

File: Misc/NEWS.d/next/Core and Builtins/2018-10-15-20-41-18.bpo-34990.NVDjdP.rst
Comment (matrixise) [reply Misc/NEWS.d/next/Core and Builtins/2018-10-15-20-41-18.bpo-34990.NVDjdP.rst]: +1

**Comment by matrixise** [comment]
@serhiy-storchaka can you continue the review with me, because I am a little bit like a chicken without head ;-) I try to understand with `0xFFFF_FFFF` ;-)

File: Lib/test/test_zipimport.py
Comment (serhiy-storchaka) [reply Lib/test/test_zipimport.py:43]: I meant that you need to update the format string.

File: Tools/scripts/checkpyc.py
Comment (serhiy-storchaka) [reply Tools/scripts/checkpyc.py:55]: Yes.

File: Tools/scripts/checkpyc.py
Comment (serhiy-storchaka) [reply Tools/scripts/checkpyc.py:57]: `mtime` is already int. But `st[ST_MTIME]` is float.

File: Misc/NEWS.d/next/Core and Builtins/2018-10-15-20-41-18.bpo-34990.NVDjdP.rst
Comment (serhiy-storchaka) [inline Misc/NEWS.d/next/Core and Builtins/2018-10-15-20-41-18.bpo-34990.NVDjdP.rst:1]: Please use punctuation.

File: Lib/test/test_zipimport.py
Comment (serhiy-storchaka) [reply Lib/test/test_zipimport.py:42]: Yes. They are roughly equivalent to `mtime = int(mtime) & 0xFFFF_FFFF` if use an unsigned integer format.

File: Lib/test/test_compileall.py
Comment (serhiy-storchaka) [inline Lib/test/test_compileall.py:205]: Unrelated change.

**Comment by brettcannon** [comment]
Due to the WIP in the title I added the DO-NOT-MERGE label.

**Comment by matrixise** [comment]
@brettcannon Thank you, I will continue to work on this issue on the next week, this week I am at the hospital with my daughter.

**Comment by bmwiedemann** [comment]
Only 19 years to go... Can we get this moving again?

**Comment by vstinner** [comment]
Ping @matrixise: they are still some comments of @serhiy-storchaka that you didn't address.

**Comment by matrixise** [comment]
@vstinner pong, ok thanks for the reminder.

**Comment by matrixise** [comment]
So, I am not serious with this bug, I am going to continue on this bug, I would like to close all my PRs.

**Comment by vstinner** [comment]
> @vstinner pong, ok thanks for the reminder.

FYI I ignore this issue since it's still tagged as "WIP". I prefer to wait until it's no longer a WIP to review it.

**Comment by brettcannon** [comment]
Removing the import team from reviewing as this is still a WIP. We can be added back when it's ready to go.

**Comment by bmwiedemann** [comment]
@matrixise any chance to get this finished this year?

**Comment by vstinner** [comment]
> @matrixise any chance to get this finished this year?

It's ok as soon as it's fixed before 2038 :-D

**Comment by bmwiedemann** [comment]
> It's ok as soon as it's fixed before 2038 :-D

I am afraid there are way too many places where software is used unchanged for a long time. Even 18 years. And of course this is not the only timebomb waiting to trigger.

**Comment by vstinner** [comment]
It was a joke. Hint: see the emoticon.

**Comment by ammaraskar** [comment]
Since Stéphane said:

> I would like to close all my PRs.

I opened up https://github.com/python/cpython/pull/19651 with the more permanent solution of using a 64-bit timestamp value as Benjamin originally suggested in their PEP 552 PR.

**Comment by ammaraskar** [comment]
I pinged @matrixise about this and they said I can feel free to take this on since they're too busy to move forward with it. https://github.com/python/cpython/pull/19708 is a continuation of this PR to make `compileall` use an unsigned int.

It should be possible to compare the code for both the solutions now.

**Comment by tirkarthi** [comment]
I guess this can be close since https://github.com/python/cpython/pull/19708 is merged

**Comment by encukou** [comment]
Indeed. Thanks for noticing.

**Pull request closed without merging**
